File: Palace/Book/UI/TPPBookDetailDownloadFailedView.m
@import PureLayout;

#import "Palace-Swift.h"

#import "TPPConfiguration.h"
#import "TPPLinearView.h"
#import "UIView+TPPViewAdditions.h"
#import "UIFont+TPPSystemFontOverride.h"
#import "TPPBookDetailDownloadFailedView.h"

@interface TPPBookDetailDownloadFailedView ()

@property (nonatomic) TPPRoundedButton *cancelButton;
@property (nonatomic) TPPLinearView *cancelTryAgainLinearView;
@property (nonatomic) UILabel *messageLabel;
@property (nonatomic) TPPRoundedButton *tryAgainButton;

@end

@implementation TPPBookDetailDownloadFailedView

- (instancetype)init
{
  self = [super init];
  if(!self) return nil;
  
  self.backgroundColor = [TPPConfiguration mainColor];
  
  self.messageLabel = [[UILabel alloc] init];
  self.messageLabel.font = [UIFont customFontForTextStyle:UIFontTextStyleBody];
  self.messageLabel.textAlignment = NSTextAlignmentCenter;
  self.messageLabel.textColor = [TPPConfiguration backgroundColor];
  self.messageLabel.text = NSLocalizedString(@"The download could not be completed.\nScroll down to 'View Issues' to see details.", nil);
  self.messageLabel.numberOfLines = 0;
  [self addSubview:self.messageLabel];
  [self.messageLabel autoPinEdgesToSuperviewEdges];
  
  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(didChangePreferredContentSize)
                                               name:UIContentSizeCategoryDidChangeNotification
                                             object:nil];
  
  return self;
}

- (void)dealloc
{
  [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)didChangePreferredContentSize
{
  self.messageLabel.font = [UIFont customFontForTextStyle:UIFontTextStyleBody];
}

- (void)configureFailMessageWithProblemDocument:(TPPProblemDocument *)problemDoc {
  if (problemDoc != nil) {
    self.messageLabel.text = NSLocalizedString(@"The download could not be completed.\nScroll down to 'View Issues' to see details.", nil);
  } else {
    self.messageLabel.text = NSLocalizedString(@"The download could not be completed.", nil);
  }
}

@end

File: Palace/Book/UI/TPPBookDetailDownloadFailedView.swift
import UIKit
import PureLayout

@objc class TPPBookDetailDownloadFailedView: UIView {

  private let messageLabel = UILabel()

  @objc init() {
    super.init(frame: .zero)
    setup()
  }

  required init?(coder: NSCoder) {
    super.init(coder: coder)
    setup()
  }

  private func setup() {
    backgroundColor = TPPConfiguration.mainColor()

    messageLabel.font = UIFont.customFont(forTextStyle: .body)
    messageLabel.textAlignment = .center
    messageLabel.textColor = TPPConfiguration.backgroundColor()
    messageLabel.text = NSLocalizedString("The download could not be completed.\nScroll down to 'View Issues' to see details.", comment: "")
    messageLabel.numberOfLines = 0
    addSubview(messageLabel)
    messageLabel.autoPinEdgesToSuperviewEdges()

    NotificationCenter.default.addObserver(
      self,
      selector: #selector(didChangePreferredContentSize),
      name: UIContentSizeCategory.didChangeNotification,
      object: nil
    )
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
  }

  @objc private func didChangePreferredContentSize() {
    messageLabel.font = UIFont.customFont(forTextStyle: .body)
  }

  @objc func configureFailMessage(with problemDoc: TPPProblemDocument?) {
    if problemDoc != nil {
      messageLabel.text = NSLocalizedString("The download could not be completed.\nScroll down to 'View Issues' to see details.", comment: "")
    } else {
      messageLabel.text = NSLocalizedString("The download could not be completed.", comment: "")
    }
  }
}
